Add `fill_with_function`

jsgf opened the pull request:
Add a `fill_with_function` where the replacement is derived by calling a function. This is much more general than replacing from a HashMap as it allows the substitutions to be computed directly from the key. It also allows things like tracking which replacements were actually used to do a substitution.

That function returns `Option<Cow<str>>` so it can either return a borrowed value or compute an owned `String`. This is always "strict" since the function can choose to either return a value (default or otherwise) or cause substitution to fail by returning `None`.

All the other fill_with_* functions are reimplemented in terms of `fill_with_function`.

jsgf committed:
Add `fill_with_function`

This is a general-purpose template replacement which allows an arbitary
key->value mapping. All the other fill_with_ functions are implemented
in terms of it.

bernardoamc commented:
Thank you for the submission @jsgf, this is a really neat idea! Do you mind also updating the README with a short description of the new functionality? :)

I should be able to push a new version of the crate tonight or tomorrow. 🙏

jsgf committed:
Doc cleanups and fixes

Several changes:
- Update README
- update documentation
- fix fill_with_struct* functions

jsgf commented:
OK I updated the readme, fixed doc comments (added examples in some cases) and also fixed the fill_with_struct* functions which I broke without noticing.

bernardoamc commented:
Thank you @jsgf! I will merge all of the pending PRs today and release a new version soon. I appreciate you taking time to improve the codebase and also document things. :)

bernardoamc commented:
Thank you @jsgf, version `0.5.0` is published. 🎉
